src/wic/api/pythonapi.py

```diff
@@ -157,11 +157,21 @@ def _is_link(s: str) -> bool:
 
 
 @singledispatch
-def _value_str(val: Any) -> Union[str, bool]:
+def _value_str(val: Any) -> Union[str, bool, int, float]:
```

Comment:
This code is unconditionally converting the input to a string, so the return type should just be str.

Perhaps more importantly, why are we creating one-line wrappers for python standard library functions?

Reply:
It is not

Reply:
The problem is that the docstring does not match the implementation.
>     """Return value of input as str."""
>     return str(val)
These two lines are consistent, but adding @singledispatch with @_value_str.register to the other implementations actually causes this to be the identity function on the whitelisted types, and the str() function otherwise.

So why are we doing that?

Reply:
What I'm trying to say is that the one and only call site is in the _yml function
`inp.name: _value_str(inp.value)`
where it appears that you are using the shorthand `name: type` syntax for specifying the **formal** parameters.
However, the _value_str function appears to operate on **arguments**.

https://en.wikipedia.org/wiki/Parameter_(computer_programming)

Reply:
IMO, the function name and the purpose does not align. We should either rename this function to something that can be helpful to infer its purpose. Also, from the one use of this function, it appears to me that it takes an input of some allowed type  and returns it as the input type (without any type conversion). So, `_value_str`  is misleading here.

Reply:
That is not the purpose of the function.
Its purpose is to help in the generation of a dictionary that can be exported by the `yaml` library to a YAML file. 
You are right that a better name - and an updated docstring - would make things better.
I will rename it and update the docstring.

Reply:
Upon closer inspection, it appears that
`inp.name: _value_str(inp.value)`
is being used to generate [inline inputs](https://workflow-inference-compiler.readthedocs.io/en/latest/userguide.html#inline-inputs), i.e. arguments.

The confusion is that there is also a shorthand `name: type` notation within the inputs: tag (for both CWL Workflow class and CWL CommandLineTool class) which would be parameters.

So let's change the name to something with inline_inputs in it, and let's put the URL to the feature somewhere. Your implementation should also (eventually, not necessarily a blocker for this PR) support all primitive CWL types.

Moreover, note that the inline inputs feature is not limited to CWL primitive types. It supports the full algebra of data types (e.g. nested lists, dicts, etc). (For certain Tools it also supports custom input schemas. See the gromacs mdp options example here https://workflow-inference-compiler.readthedocs.io/en/latest/installguide.html#intellisense-code-completion)

So this code is making assumptions and restrictions that are not required for the underlying yml syntax. This will need to be fixed in future PRs.

Reply:
> Upon closer inspection, it appears that `inp.name: _value_str(inp.value)` is being used to generate [inline inputs](https://workflow-inference-compiler.readthedocs.io/en/latest/userguide.html#inline-inputs), i.e. arguments.
> 
> The confusion is that there is also a shorthand `name: type` notation within the inputs: tag (for both CWL Workflow class and CWL CommandLineTool class) which would be parameters.
> 
> So let's change the name to something with inline_inputs in it, and let's put the URL to the feature somewhere. Your implementation should also (eventually, not necessarily a blocker for this PR) support all primitive CWL types.
> 
> Moreover, note that the inline inputs feature is not limited to CWL primitive types. It supports the full algebra of data types (e.g. nested lists, dicts, etc). (For certain Tools it also supports custom input schemas. See the gromacs mdp options example here https://workflow-inference-compiler.readthedocs.io/en/latest/installguide.html#intellisense-code-completion)
> 
> So this code is making assumptions and restrictions that are not required for the underlying yml syntax. This will need to be fixed in future PRs.

This function only referes to things that will be serialized by `yaml`

Reply:
Using this definition for echo
```
cwlVersion: v1.2

class: CommandLineTool

baseCommand: echo

inputs:
  message:
    type: Any  # NOTE: This requires v1.2 above
    inputBinding:
      position: 1

outputs:
  stdout:
    type: File
    outputBinding:
      glob: stdout

stdout: stdout
```
Consider the following **wic** yaml file echo_types.yml:
```
steps:
- echo:
    in:
      message: Hello World
- echo:
    in:
      message: 1
- echo:
    in:
      message: 2.0
# There are many more types than just string, int, and float:
- echo:
    in:
      message: [a, list, of, strings]
- echo:
    in:
      message: [[2D, array], [of, strings]]
- echo:
    in:
      message: ['nested', {'data': ['structures']}]
- echo:
    in:
      message: ['arbitrary': {'json': {'blob': [1, 2.0, {3.14: {}}]}}]
# i.e. there is an entire algebra of data types!
```
```
wic --yaml echo_types.yml --ignore_validation_errors --run_local
...
Success! Output files should be in outdir/
```
Notice that I bolded **wic** yaml above, because the wic python API isn't supposed to just generate any random yaml, it is supposed to generate yaml according to the wic schema. Please [readthedocs](https://workflow-inference-compiler.readthedocs.io/en/latest/validation.html#the-wic-schema).

Reply:
@camilovelezr Since this PR is a hotfix for one of our internal users, we can merge it as-is.

But for future reference, my point is that the _value_str / _yaml_value function is an (incomplete!) identity function. So rather than try to extend a not-quite-identity function to cover more types, we should just delete it.